src/main/java/desafio/votacao/service/SessaoVotacao/SessaoVotacaoServiceImpl.java

```diff
@@ -0,0 +1,95 @@
+
+import java.time.LocalTime;
+import java.util.Optional;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import desafio.votacao.dto.Voto.RequestVotoDto;
+import desafio.votacao.enums.Resultado;
+import desafio.votacao.enums.Situacao;
+import desafio.votacao.enums.TipoVoto;
+import desafio.votacao.exception.NotFoundException;
+import desafio.votacao.model.Pauta;
+import desafio.votacao.model.SessaoVotacao;
+import desafio.votacao.repository.SessaoVotacaoRepository;
+
+@Service
+public class SessaoVotacaoServiceImpl implements SessaoVotacaoService {
+    
+    @Autowired
+    SessaoVotacaoRepository repository;
+
+    //Criada uma instância do Scheduled com um pool de threads de tamanho 1, para especificar que apenas uma tarefa será executada por vez neste executor. 
+    private ScheduledExecutorService scheduler = Executors.newScheduledThreadPool(1);
+
+
+    @Override
+    public void abrirSessaoVotacao(int tempo, Pauta pauta){
+
+        SessaoVotacao sessaoVotacao = SessaoVotacao.builder()
+                                                    .ativa(true)
+                                                    .pauta(pauta)
+                                                    .tempoInicioSessao(LocalTime.now().withNano(0))
+                                                    .tempoFimSessao(LocalTime.now().plusMinutes(tempo).withNano(0))
+                                                    .situacao(Situacao.ABERTA)
+                                                    .build();  
+
+        repository.save(sessaoVotacao);
+
+        // Agenda a tarefa para chamar a função a cada minuto
+        scheduler.scheduleAtFixedRate(() -> verificaSeTempoSessaoExpirou(sessaoVotacao), 1, 1, TimeUnit.MINUTES);
+    }
+
+    @Override
+    public Optional<SessaoVotacao> buscarSessaoVotacao(Long id){
+        Optional<SessaoVotacao> sessaoVotacao = repository.findById(id);
+        
+        if(sessaoVotacao.isEmpty()){
+            throw new NotFoundException("Não foi possivel encontrar essa sessão de votação");
+        }
+
+        return sessaoVotacao;
+    }
+
+    @Override
+    public void contabilizarVotoNaSessao(Long id, RequestVotoDto dto){
+        SessaoVotacao sessaoVotacao = buscarSessaoVotacao(id).get();
+
+        if ((dto.tipo() == TipoVoto.SIM)) {
+            sessaoVotacao.setVotosSim(sessaoVotacao.getVotosSim() + 1);
+        }else if (dto.tipo() == TipoVoto.NAO) {
+            sessaoVotacao.setVotosNao(sessaoVotacao.getVotosNao() + 1);
+        }
+
+        repository.save(sessaoVotacao);
+    }
+
+    @Override
+    public void verificaSeTempoSessaoExpirou(SessaoVotacao sessaoVotacao ){
+        LocalTime agora = LocalTime.now().withNano(0);
+
+        if (agora.equals(sessaoVotacao.getTempoFimSessao())) {
```

Comment:
Supondo que uma sessão comece as 12:00:00 e tenha 1 hora e 30 segundos de duração, o fim dessa sessão será as 13:00:30.

Pelo que entendi o scheduler vai fazer uma verificação a cada minuto então metodo`verificaSeTempoSessaoExpirou`, será executado as 13:01:00, o que torna:

`if (agora.equals(sessaoVotacao.getTempoFimSessao()))`

sempre falso para esse caso!

Reply:
De fato, ocorreria exatamente como você mencionou. Para resolver esse caso específico, eu mudaria a unidade de tempo do scheduler para TimeUnit.SECONDS, de forma a realizar a verificação a cada segundo.

Neste exercício, optei por configurar a verificação a cada minuto, permitindo no DTO apenas números inteiros, evitando assim tempos em segundos.

Se for uma melhor prática permitir tempos em segundos, posso fazer a alteração sem problemas.